src/reconstruction/gapfill.jl

```diff
@@ -0,0 +1,125 @@
+    modifications=[], 
+    weights = fill(1.0, length(universal_reactions)),
+)
+    # constraints from model to be gap filled
+    S_model = stoichiometry(model)
+    mids = metabolites(model)
+    lbs, ubs = bounds(model)
+
+    # set objective bounds
+    obj_idx = first(indexin([objective_id_bounds[1]], reactions(model)))
+    lbs[obj_idx] = objective_id_bounds[2]
+    ubs[obj_idx] = objective_id_bounds[3]
+
+    # constraints from universal reactions that can fill gaps
+    n_universal_reactions = length(universal_reactions)
+    n_rxns_model = size(S_model, 2)
+    S_universal, lbs_universal, ubs_universal = COBREXA._universal_stoichiometry(universal_reactions, mids)
+
+    # optimization problem is a MILP 
+    n_vars = size(S_model, 2) + size(S_universal, 2)
+    S = [
+            [
+                S_model
+                spzeros(size(S_universal, 1) - size(S_model, 1), size(S_model, 2))
+            ] S_universal
+    ]
+    bal = [
+        balance(model)
+        spzeros(size(S_universal, 1)-size(S_model, 1))
+    ]
+
+    opt_model = JuMP.Model(optimizer)
+    @variable(opt_model, x[1:n_vars]) # fluxes
+    @variable(opt_model, y[1:n_universal_reactions], Bin) # indicators
+    @constraint(opt_model, mb, S * x .== bal) # mass balance
+    # flux bounds of model
+    @constraint(opt_model, lbs, lbs .<= x[1:n_rxns_model])
+    @constraint(opt_model, ubs, x[1:n_rxns_model] .<= ubs)
+    # flux bounds of universal reactions with indicators
+    @constraint(opt_model, lbs_universal, lbs_universal .* y .<= x[(n_rxns_model+1):end]) 
+    @constraint(opt_model, ubs_universal, x[(n_rxns_model+1):end] .<= ubs_universal .* y) 
+
+    @objective(opt_model, Min, sum(weights .* y))
+
+    for mod in modifications
+        mod(opt_model, model)
+    end
+
+    optimize!(opt_model)
+
+    [universal_reactions[idx].id for idx in findall(value.(y) .> 0)]
+end
+
+"""
+    _universal_stoichiometry(
+        universal_reactions::Vector{Reaction},
+        metabolite_id_order,
+    )
+
+A helper function that constructs the stoichiometric matrix of a set of
+`universal_reactions`. The order of the metabolites is determined with
+`metabolite_id_order`, so that this stoichiometric matrix can be combined with
+another one.
+"""
+function _universal_stoichiometry(
+    universal_reactions::Vector{Reaction},
+    metabolite_id_order,
+)
+    rows = Int[]
+    cols = Int[]
+    vals = Int[]
+    lbs = zeros(length(universal_reactions))
+    ubs = zeros(length(universal_reactions))
+    met_id_order_lu = Dict(zip(metabolite_id_order, 1:length(metabolite_id_order)))
+    n_midxs = length(met_id_order_lu)
+
+    for (col, rxn) in enumerate(universal_reactions)
+        for (mid, stoich) in rxn.metabolites 
+            if !haskey(met_id_order_lu, mid)
+                n_midxs += 1
+                met_id_order_lu[mid] = n_midxs
+            end
+            push!(rows, met_id_order_lu[mid])
+            push!(cols, col)
+            push!(vals, stoich)
```

Comment:
unnecessarily mutable, better use sparse() right away. (Similarly as we do with the gecko matrices)

roughly:
```julia
sparse(
  (ridx for (ridx, rxn) in enumerate(uni_rxns) for (mid, stoi) in rxn.metabolites if haskey(met_id_order_lu, mid)),
  (met_id_order_lu[mid] for (ridx, rxn) in enumerate(uni_rxns) for (mid, stoi) in rxn.metabolites if haskey(met_id_order_lu, mid)),
  (stoi for (ridx, rxn) in enumerate(uni_rxns) for (mid, stoi) in rxn.metabolites if haskey(met_id_order_lu, mid)),
  length(metabolite_id_order),
  length(universal_reactions),
)
```

the cycling is a bit verbose but I don't know if we have machinery to actually share the code better than macros.

Note there are no `[ ]` around the vectors, sparse() can work with ranges.

Reply:
(re code sharing -- actually yeah you can make a function that runs the generator and parametrize it with a function that can choose between ridx, mid and stoi)

Reply:
yyyeeaaahhh but the mutable one is so much easier to read...

Reply:
let's talk about this more tomorrow